File: src/lib/stores/tasks.ts
import { writable } from 'svelte/store';
import { invoke } from '$lib/tauri';

export interface Task {
  id: string;
  name: string;
  quadrant: number;
  completed: boolean;
  position: number;
  device_id: string;
  updated_at: number;
}

export const q1Tasks = writable<Task[]>([]);
export const q2Tasks = writable<Task[]>([]);

export async function loadTasks() {
  const [q1, q2] = await Promise.all([
    invoke<Task[]>('cmd_list_tasks', { quadrant: 1 }),
    invoke<Task[]>('cmd_list_tasks', { quadrant: 2 }),
  ]);
  q1Tasks.set(q1);
  q2Tasks.set(q2);
}

export async function addTask(name: string, quadrant: number) {
  await invoke('cmd_create_task', { name, quadrant });
  await loadTasks();
}

export async function completeTask(id: string) {
  await invoke('cmd_complete_task', { id });
  await loadTasks();
}

export async function deleteTask(id: string) {
  await invoke('cmd_delete_task', { id });
  await loadTasks();
}

export async function reorderTasks(orderedIds: string[]) {
  await invoke('cmd_update_task_positions', { ordered_ids: orderedIds });
  await loadTasks();
}

File: src-tauri/src/commands/tasks.rs
use crate::{tasks, AppState};
use tauri::State;

#[tauri::command]
pub async fn cmd_create_task(
    state: State<'_, AppState>,
    name: String,
    quadrant: i64,
) -> Result<tasks::Task, String> {
    tasks::create_task(&state.db, &name, quadrant, &state.device_id)
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn cmd_list_tasks(
    state: State<'_, AppState>,
    quadrant: i64,
) -> Result<Vec<tasks::Task>, String> {
    tasks::list_tasks(&state.db, quadrant)
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn cmd_complete_task(
    state: State<'_, AppState>,
    id: String,
) -> Result<(), String> {
    tasks::complete_task(&state.db, &id)
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn cmd_delete_task(
    state: State<'_, AppState>,
    id: String,
) -> Result<(), String> {
    tasks::delete_task(&state.db, &id)
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn cmd_update_task_positions(
    state: State<'_, AppState>,
    ordered_ids: Vec<String>,
) -> Result<(), String> {
    tasks::update_task_positions(&state.db, &ordered_ids)
        .await
        .map_err(|e| e.to_string())
}
